cmd/promoted-image-governor/main.go

```diff
@@ -303,6 +309,62 @@ func isMirroredFromOCP(tag api.ImageStreamTagReference, refs []ImageStreamRef) b
 	return false
 }
 
+func deleteTagsOnBuildFarm(ctx context.Context, appCIClient ctrlruntimeclient.Client, buildClusterClients map[string]ctrlruntimeclient.Client, imageStreamsWithPromotedTags map[ctrlruntimeclient.ObjectKey]interface{}, dryRun bool) error {
+	var errs []error
+	for streamKey := range imageStreamsWithPromotedTags {
+		for cluster, client := range buildClusterClients {
+			imageStream := &imagev1.ImageStream{}
+			if err := client.Get(ctx, streamKey, imageStream); err != nil {
+				if !kerrors.IsNotFound(err) {
+					errs = append(errs, fmt.Errorf("could not get image stream %s in namespace %s on cluster %s: %w", streamKey.Name, streamKey.Namespace, cluster, err))
+				} else {
+					logrus.WithField("cluster", cluster).WithField("streamKey", streamKey).Debug("image stream not found")
+				}
+				continue
+			}
+			for _, tag := range imageStream.Status.Tags {
```

Comment:
Instead of fanning out and doing N API calls per tag, why don't we just fetch the stream, then process the list of tags client-side?

Reply:
This is why it took so much time on each run.
I will change.